### DISCUSSION TITLE
Equivalent to Picasso's LoadedFrom?

### DISCUSSION THREAD

**Discussion by rogererill** [coil-kt/coil#965] [Q&A] [answered]
Hi! I'm in the process of migrating code from Picasso to Coil :) and so far everything's good but now I'm stuck into this problem and I don't know if it has a solution. When working with `Target`, in Picasso I had:

    val picassoTarget = object : Target {
        override fun onPrepareLoad(placeHolderDrawable: Drawable?) {
            // no-op
        }

        override fun onBitmapFailed(e: Exception, errorDrawable: Drawable) {
            // no-op
        }

        override fun onBitmapLoaded(bitmap: Bitmap, from: Picasso.LoadedFrom) {
            if (from == LoadedFrom.MEMORY) {
               // Do something
            }
        }
    }
    requestCreator.into(picassoTarget)

But now, the equivalent code in Coil is pretty similar but without a way to know where the bitmap came from:

        val coilTarget = object : Target {
            override fun onError(error: Drawable?) {
               // no-op
            }

            override fun onStart(placeholder: Drawable?) {
                // no-op
            }

            override fun onSuccess(result: Drawable) {
                // how to replicate logic of Picasso's onBitmapLoaded?
            }
        }

Is there any way to achieve this? Thanks in advance for your time :)

**Comment by colinrtwhite** [comment, marked as answer]
In Coil `LoadedFrom` == `DataSource` and you can access it in the listener:

```kotlin
listener(onSuccess = { request, metadata ->
    if (metadata.dataSource == DataSource.MEMORY) {
           // Do something
    }
})
```

Note: Coil has a `DataSource.MEMORY_CACHE` enum which is slightly different than `DataSource.MEMORY`. Check out the docs.
